### PR TITLE
enable dry-run option

### PR DESCRIPTION
**Pull request by karthiktv006**

### CODE REVIEW

**Comment by jbpros** [comment]
Thank you. Can you please add/adapt e2e and unit tests for this?

**Comment by karthiktv006** [comment]
@jbpros I have written few unit test for it, please let me know if some changes has to be done.

**Comment by olore** [comment]
cool!

**Comment by jlin412** [comment]
@jbpros All the tests are passing.  Are we waiting for anything else before merge?

File: lib/cucumber/runtime/ast_tree_walker.js [hunk 1]
```diff
@@ -288,13 +292,15 @@ function AstTreeWalker(features, supportCodeLibrary, listeners, strictMode) {
     },
 
     processStep: function processStep(step, callback) {
-      if (self.isStepUndefined(step)) {
+      if(self.isDryRun()) {
+        self.skipStep(step, callback);
+      } else if (self.isStepUndefined(step)) {
```
Comment (charlierudolph) [inline lib/cucumber/runtime/ast_tree_walker.js]: dry run shouldn't skip undefined steps if we are trying to follow cucumber's [example](http://www.relishapp.com/cucumber/cucumber/docs/cli/dry-run). I'd suggest adding some feature tests for this.

File: lib/cucumber/runtime/ast_tree_walker.js [hunk 2]
```diff
@@ -288,13 +292,15 @@ function AstTreeWalker(features, supportCodeLibrary, listeners, strictMode) {
     },
 
     processStep: function processStep(step, callback) {
-      if (self.isStepUndefined(step)) {
+      if(self.isDryRun()) {
```
Comment (charlierudolph) [inline lib/cucumber/runtime/ast_tree_walker.js]: There should be a space between the if and the paren in keeping with the rest of the repo

File: lib/cucumber/runtime/ast_tree_walker.js [hunk 3]
```diff
@@ -288,13 +292,15 @@ function AstTreeWalker(features, supportCodeLibrary, listeners, strictMode) {
...
+        self.skipStep(step, callback);
+      } else if (self.isStepUndefined(step)) {
         self.witnessUndefinedStep();
         self.skipUndefinedStep(step, callback);
       } else if (self.isSkippingSteps()) {
         self.skipStep(step, callback);
       } else {
-        self.executeStep(step, callback);
+          self.executeStep(step, callback);
```
Comment (charlierudolph) [inline lib/cucumber/runtime/ast_tree_walker.js]: this change should probably be reverted

File: lib/cucumber/runtime.js [hunk 4]
```diff
@@ -7,15 +7,20 @@ function Runtime(configuration) {
...
       try {
         isStrictRequested    = configuration.isStrictRequested();
       } catch(e) {
         isStrictRequested    = false;
       }
+      try {
+        isDryRunRequested    = configuration.isDryRunRequested();
+      } catch(e) {
+        isDryRunRequested    = false;
+      }
```
Comment (charlierudolph) [inline lib/cucumber/runtime.js]: I don't see any reason for the try/catch. Looks like you were just following the style of isStrictRequested, but I believe that can be removed too.

**Comment by jbpros** [comment]
This is new user-facing behaviour. It deserves a few e2e tests (i.e. cucumber scenarios :)).

File: lib/cucumber/runtime.js [hunk 5]
```diff
@@ -7,15 +7,12 @@ function Runtime(configuration) {
...
       if (typeof(callback) !== 'function')
         throw new Error(Runtime.START_MISSING_CALLBACK_ERROR);
-      var isStrictRequested;
-      try {
-        isStrictRequested    = configuration.isStrictRequested();
-      } catch(e) {
-        isStrictRequested    = false;
-      }
+
+      var isStrictRequested  = configuration.isStrictRequested && configuration.isStrictRequested();
```
Comment (karthiktv006) [inline lib/cucumber/runtime.js]: Fixes #325

**Comment by karthiktv006** [comment]
@jbpros @charlierudolph - Please take a look; e2e is added and some changes are made based on the comments.

File: lib/cucumber/runtime.js (hunk 5 above)
Comment (charlierudolph) [inline lib/cucumber/runtime.js]: Please revert this change. Its unrelated and I'd prefer to have it dealt with separately

File: features/dryrun_mode.feature [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+  Using the `--dry-run` or `-d` flag gives you a way to quickly scan your features without actually running them.
+
+  Background:
+    Given a file named "features/a.feature" with:
+    """
+    Feature: Missing
+      Scenario: Missing
+        Given this step passes
+    """
```
Comment (charlierudolph) [inline features/dryrun_mode.feature]: Please indent all the doc strings by two spaces.

File: features/dryrun_mode.feature [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    1 step (1 undefined)
+
+    You can implement step definitions for undefined steps with these snippets:
+
+    this.Given(/^this step passes$/, function (callback) {
+      // Write code here that turns the phrase above into concrete actions
+      callback.pending();
+    });
+    """
+    And the exit status should be 1
```
Comment (charlierudolph) [inline features/dryrun_mode.feature]: I don't think this test case is needed.
strict has its own test case that -S and --strict is synonymous.

File: features/dryrun_mode.feature [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    When I run cucumber.js with `-f progress features/a.feature --dry-run`
+    Then it outputs this text:
+    """
+    -
+
+    1 scenario (1 skipped)
+    1 step (1 skipped)
+    """
+    And the exit status should be 0
```
Comment (charlierudolph) [inline features/dryrun_mode.feature]: I don't think this is needed. The steps aren't actually run so there is no way to tell the callback is pending

File: features/dryrun_mode.feature [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    -
+
+    1 scenario (1 skipped)
+    1 step (1 skipped)
+
+    """
+    And the exit status should be 0
+
+  Scenario: Skipped scenario with passing step with --dry-run --strict
```
Comment (charlierudolph) [inline features/dryrun_mode.feature]: can you please use `defined` instead of `implemented` or `passing` when describing the step?

**Comment by karthiktv006** [comment]
@charlierudolph changes are done.

File: lib/cucumber/runtime/ast_tree_walker.js [hunk 10]
```diff
@@ -196,6 +196,10 @@ function AstTreeWalker(features, supportCodeLibrary, listeners, strictMode) {
       return world;
     },
 
+    isDryRun: function isDryRun() {
+      return dryRun;
+    },
```
Comment (charlierudolph) [inline lib/cucumber/runtime/ast_tree_walker.js]: This function is unnecessary just use dryRun directly.

Looks like this was useful in tests. Instead just create another tree walker with a different dryRun value.

**Comment by charlierudolph** [comment]
Awesome. Just one last thing, then this is good to merge. Thank you for all your hard work!

**Comment by karthiktv006** [comment]
@charlierudolph isDryRun is removed.

**Comment by charlierudolph** [comment]
Thanks for your contribution @karthiktv006
